src/pybind/mgr/volumes/fs/nfs.py

```diff
@@ -0,0 +1,373 @@
+
+    def check_fs(self, fs_name):
+        fs_map = self.mgr.get('fs_map')
+        return fs_name in [fs['mdsmap']['fs_name'] for fs in fs_map['filesystems']]
+
+    def check_pseudo_path(self, pseudo_path):
+        for ex in self.exports[self.rados_namespace]:
+            if ex.pseudo == pseudo_path:
+                return True
+        return False
+
+    def _create_user_key(self, entity):
+        osd_cap = 'allow rw pool={} namespace={}, allow rw tag cephfs data=a'.format(
+                self.rados_pool, self.rados_namespace)
+        ret, out, err = self.mgr.mon_command({
+            'prefix': 'auth get-or-create',
+            'entity': 'client.{}'.format(entity),
+            'caps' : ['mon', 'allow r', 'osd', osd_cap, 'mds', 'allow rw path=/'],
+            'format': 'json',
+            })
+
+        if ret!= 0:
+            return ret, err
+
+        json_res = json.loads(out)
+        log.info("Export user is {}".format(json_res[0]['entity']))
+
+        return json_res[0]['entity'], json_res[0]['key']
+
+    def format_path(self, path):
+        if path is not None:
+            path = path.strip()
+            if len(path) > 1 and path[-1] == '/':
+                path = path[:-1]
+        return path
+
+    def _gen_export_id(self):
+        exports = sorted([ex.export_id for ex in self.exports[self.rados_namespace]])
+        nid = 1
+        for e_id in exports:
+            if e_id == nid:
+                nid += 1
+            else:
+                break
+        return nid
+
+    def _write_raw_config(self, conf_block, obj, append=False):
+        raw_config = GaneshaConfParser.write_conf(conf_block)
+        with self.mgr.rados.open_ioctx(self.rados_pool) as ioctx:
+            if self.rados_namespace:
+                ioctx.set_namespace(self.rados_namespace)
+            if append:
+                ioctx.append(obj, raw_config.encode('utf-8'))
+            else:
+                ioctx.write_full(obj, raw_config.encode('utf-8'))
+            log.debug(
+                    "write configuration into rados object %s/%s/%s:\n%s",
+                    self.rados_pool, self.rados_namespace, obj, raw_config)
+
+    def _update_common_conf(self, ex_id):
+        common_conf = 'conf-nfs'
+        conf_blocks = {
+                'block_name': '%url',
+                'value': self.make_rados_url(
+                'export-{}'.format(ex_id))
+                }
+        self._write_raw_config(conf_blocks, common_conf, True)
+
+    def _save_export(self, export):
+        self.exports[self.rados_namespace].append(export)
+        conf_block = export.to_export_block()
+        self._write_raw_config(conf_block, "export-{}".format(export.export_id))
+        self._update_common_conf(export.export_id)
+
+    def create_export(self, export_type, fs_name, pseudo_path, read_only, path, cluster_id):
+        if export_type != 'fs':
+            return -errno.EINVAL,"", "Invalid FSAL. Only CephFS is supported."
```

Comment:
```suggestion
            return -errno.EINVAL, "", "Invalid export_type: {export_type}"
```
maybe make this more explicit?  `Invalid FSAL` is somewhat confusing ..

Reply:
Updated.